docs: Offline Data Preprocessing documentation (#502)
foundation-model-stack/fms-hf-tuning: docs/offline-data-preprocessing.md

# Offline Data Preprocessing

> Our library provides a [script](../scripts/offline_data_processing.py) that allows users to perform standalone data preprocessing, independent of `tuning/training`. This script enables users to process raw datasets, apply basic/advance data preprocessing, and save the train and validation datasets in `Parquet` format inside the specified `output_dir`. When the `--num_dataset_shards` argument is specified, the datasets are divided and saved into multiple shards.

Reviewer: nit:
```suggestion
Our library provides a [script](../scripts/offline_data_processing.py) that allows users to perform standalone data preprocessing, independent of tuning/training. This script enables users to process raw datasets, apply basic/advanced data preprocessing, and save the train and validation datasets in Parquet format inside the specified `output_dir`. When the `--num_dataset_shards` argument is specified, the datasets are divided and saved into multiple shards.
```

> A data configuration YAML file can be used to pass configuration parameters to this script. Example data config file:

Reviewer 2: ```suggestion
Users can pass any data config to this script and the goal of this script is to take data config and generate a datset which can be used directly while training without any need of online processing. As an example see this data config below
```

```yaml
dataprocessor:
    type: default
    sampling_stopping_strategy: first_exhausted
    seed: 66
datasets:
  - name: dataset_1
    data_paths:
      - tests/artifacts/testdata/jsonl/twitter_complaints_input_output.jsonl
    data_handlers:
      - name: tokenize_and_apply_input_masking
        arguments:
          remove_columns: all
          batched: false
          fn_kwargs:
            input_field_name: input
            output_field_name: output
```

After preparing the data configuration YAML file, run the script with the following example command to perform offline data preprocessing:   

```
python scripts/offline_data_processing.py \
--data_config_path  /path/to/data_config.yaml \
--model_name_or_path "model_name"  \
--max_seq_length 4096 \
--output_dir /path/to/output/directory  \
--log_level info \
--num_dataset_shards 3
```

> Additionally, once the offline data processing is complete, users can leverage the shards stored in `output_dir` for tuning, provided they find the sharded datasets beneficial for training.

Reviewer: Even though this is shown below, I think it's good to reiterate here as well
```suggestion
Additionally, once the offline data processing is complete, users can leverage the shards stored in `output_dir` for tuning by passing it through the `--training_data_path` flag, provided they find the sharded datasets beneficial for training.
```

## Example Usage

> ### Applying Chat Template

Reviewer 2: ```suggestion
Below is a sample use case of how to train post offline processing a dataset where chat template is applied and eventual training is run using the offline processed dataset
```
